### PR TITLE
core: import routes directly from RailJSON

### PR DESCRIPTION
**Pull request by multun**
Load routes before signals (improves conditionalParameters load)

Also:
* LoadingGaugeConstraint is now a data class (no more interface) as it is compared
* Add benchmarking to ValidateInfra (not targeting `dev` branch), so we have a ratio of the new load vs legacy one:
  :de: Germany 27% :
 [ValidateInfra] total elapsed legacy: 19679.565198333334
 [ValidateInfra] total elapsed new:     5444.932856

  :fr: France 70% :
 [ValidateInfra] total elapsed legacy: 9922.474726666665
 [ValidateInfra] total elapsed new:     6979.596592

  :mag: small_infra 36% :
 [ValidateInfra] total elapsed legacy: 36.689171333333334
 [ValidateInfra] total elapsed new:     13.540263
* Compare checks on resulting infra (still ✅ on France and Germany)
* Remove/clean stitching between legacy load and new load in "fromRJS" load
* Introduce more consistent structure to link detectors, chunks' boundaries and track sections

Fix: https://github.com/OpenRailAssociation/osrd/issues/7055

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 1]
```diff
@@ -359,24 +329,38 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
-                )
+        val chunkBoundariesOffsets = mutableOffsetArrayListOf<TrackSection>(Offset(0.meters))
+        val chunkBoundariesDetectors = mutableListOf<List<String>?>(null)
+
+        // add all detectors
+        val trackSectionDetectors = trackSectionNameToDistanceSortedDetectors[rjsTrack.id]
+        if (trackSectionDetectors != null) {
+            for (detector in trackSectionDetectors) {
+                val detectorOffset = detector.key
+                if (detectorOffset > trackSectionLength || detectorOffset < Offset(Distance.ZERO)) {
```
Comment (Khoyo) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Offset.ZERO ?

File: core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/indexing/StaticIdx.kt [hunk 2]
```diff
@@ -36,6 +36,9 @@ value class OptStaticIdx<T>(private val data: UInt) {
             onSome(asIndex())
         }
     }
+
+    val box: StaticIdx<T>?
```
Comment (bougue-pe) [inline core/kt-osrd-utils/src/main/kotlin/fr/sncf/osrd/utils/indexing/StaticIdx.kt]: nit: Maybe a clearer name (as box/unwrap is ambiguous in the context)
```suggestion
    val orNull: StaticIdx<T>?
```

**Review by Khoyo — APPROVED** [review]
Otherwise :shipit:

File: core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt [hunk 3]
```diff
@@ -312,19 +315,24 @@ fun assertEqualSimInfra(left: RawInfra, right: RawInfra) {
...
+        compareResult(zonePathIndex) { getZonePathEntry(it) }
+        compareResult(zonePathIndex) { getZonePathExit(it) }
+        compareResult(zonePathIndex) { getZonePathLength(it) }
+        compareResult(zonePathIndex) { getZonePathChunks(it) }
+        compareResult(zonePathIndex) { getSignals(it) }
+        compareResult(zonePathIndex) { getSignalPositions(it) }
+        compareResult(zonePathIndex) { getZonePathMovableElements(it) }
+        compareResult(zonePathIndex) { getZonePathMovableElementsConfigs(it) }
+        compareResult(zonePathIndex) { getZonePathMovableElementsPositions(it) }
+    }
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt:176]: Additional possible checks:
- findZonePath(entry, exit, movable, configs)
- getSpeedLimits/Starts/Ends(route)
- getRouteReleaseZones()

The reverse maps (zone+routeNameMap + DirDetEntry/ExitToRoute) are built the same way, so it's probably OK to skip tests.

File: core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt [hunk 4]
```diff
@@ -226,6 +226,9 @@ fun assertEqualSimInfra(left: RawInfra, right: RawInfra) {
             rightTrackChunks[chunkKey] = ComparableChunk(right, c)
         }
     }
+    val leftDiff = leftTrackChunks.entries - rightTrackChunks.entries
+    val rightDiff = rightTrackChunks.entries - leftTrackChunks.entries
+    assert(leftDiff.isEmpty() && rightDiff.isEmpty())
     assert(leftTrackChunks == rightTrackChunks)
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/utils/SimInfraAdapterComparatorUtils.kt]: This last assert is rendundant (I'd keep this one only).

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 5]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
-        for (entry in
-            builder.getTrackSectionDistanceSortedChunkMap()[oldTrackSectionMap[track]]!!) {
-            chunkMap[entry.key] = entry.value
+    // parse routes
+    val routeNameIndex = mutableMapOf<String, RouteId>()
+    for (rjsRoute in rjsInfra.routes) {
+        val routeName = rjsRoute.id
+
+        // parse the entry / exit detectors
+        val routeEntryDetector = builder.getDetectorByName(rjsRoute.entryPoint.id.id)!!
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Improve error message ? (same on exit and releaseDet below)

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 6]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
+        val routeEntryDirection = rjsRoute.entryPointDirection.toDirection()
+        val routeEntry: DirDetectorId = DirStaticIdx(routeEntryDetector, routeEntryDirection)
+        val routeExit: DetectorId = builder.getDetectorByName(rjsRoute.exitPoint.id.id)!!
+
+        // parse release detectors
+        val releaseDetectors = MutableStaticIdxArraySet<Detector>(10)
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Why reserve 10?

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 7]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
         }
-        oldTrackChunkMap[track] = chunkMap
 
-        for (detector in track.detectors) {
-            oldDetectorMap[detector] = detectorNameIndex[detector.id]
+        // parse route node configs
+        val routeNodes = mutableMapOf<TrackNodeId, TrackNodeConfigId>()
+        for (rjsRouteNode in rjsRoute.switchesDirections) {
+            val node = builder.getTrackNodeByName(rjsRouteNode.key)!!
+            val config = builder.getTrackNodeConfigByName(node, rjsRouteNode.value)!!
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Improve error messages

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 8]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
-                    /* add these parameters */
-                    Map<String, String>,
-                    /* to this map, which was already passed to the builder */
-                    MutableMap<RouteId, Map<String, String>>
-                >
-            >
-        >()
+        routeNameIndex[routeName] =
+            builder.route(routeName) {
+                buildRoute(routeEntry, routeExit, routeNodes, releaseDetectors, builder)
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: How about merging it all into Builder?

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 9]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
 
     // parse signals
     for (rjsSignal in rjsInfra.signals) {
-        val trackSectionId = builder.getTrackSectionByName(rjsSignal.track)
-        val direction =
-            when (rjsSignal.direction!!) {
-                EdgeDirection.START_TO_STOP -> INCREASING
-                EdgeDirection.STOP_TO_START -> DECREASING
-            }
+        val trackSectionId = builder.getTrackSectionByName(rjsSignal.track)!!
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Improve error message

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 10]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
-            }
+    return builder.build()
+}
 
-        // TODO: remove this hack once signals are parsed after routes
-        val routeCondParams = delayedConditionalParameters[route.id] ?: continue
-        for (routeCondParam in routeCondParams) {
-            routeCondParam.second[routeId] = routeCondParam.first
-        }
+fun EdgeDirection.toDirection(): Direction {
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Shouldn't we move that to `utils/Direction.kt` instead ?

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 11]
```diff
@@ -646,195 +630,95 @@ fun adaptRawInfra(infra: SignalingInfra, rjsInfra: RJSInfra): RawInfra {
...
-    exit: DirDetectorId,
-    trackNodeMap: HashBiMap<Switch, TrackNodeId>,
-    trackNodeGroupsMap: MutableMap<Switch, Map<String, TrackNodeConfigId>>,
-    trackChunkMap: MutableMap<UndirectedTrackSection, Map<Distance, TrackChunkId>>,
+private fun RouteBuilder.buildRoute(
+    routeEntry: DirDetectorId,
+    exit: DetectorId,
+    nodeConfigs: Map<TrackNodeId, TrackNodeConfigId>,
+    releaseDetectors: StaticIdxSortedSet<Detector>,
     builder: RawInfraFromRjsBuilder
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]:
```suggestion
    infraBuilder: RawInfraFromRjsBuilder
```

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt [hunk 12]
```diff
@@ -844,107 +728,118 @@ private fun buildZonePath(
...
+                    DECREASING -> curTrackChunks[nextChunkBoundary]
+                }
+            val trackDir = curTrack.direction
+            addChunk(curChunk.withDirection(trackDir))
+            curChunkBoundary = nextChunkBoundary
+        }
+
+        val endpoint = curTrack.toEndpoint
+        val nextNode = builder.getNodeAtEndpoint(endpoint)!!
+        val nextNodeConfig = nodeConfigs[nextNode]!!
```
Comment (bougue-pe) [inline core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: Error message on invalid route?

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 13]
```diff
@@ -164,22 +164,25 @@ class PhysicalSignalBuilderImpl(
...
     }
 }
 
 private class TrackSectionBuilder(
     val name: String,
     val chunks: StaticIdxList<TrackChunk>,
     val detectors: MutableStaticIdxList<Detector> = mutableStaticIdxArrayListOf(),
-    val detectorPositions: MutableOffsetList<TrackSection> = mutableOffsetArrayListOf(),
+    // a lookup table which acts as a reverse index for detectorBoundary: for a given chunk
+    // boundary,
```
Comment (bougue-pe) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt]: nit: comment wrap

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 14]
```diff
@@ -227,6 +253,20 @@ class RawInfraFromRjsBuilder {
...
     }
 
+    fun getNextTrackSection(
+        curTrack: DirTrackSectionId,
+        config: TrackNodeConfigId
+    ): DirTrackSectionId? {
+        val nextNode = getNodeAtEndpoint(curTrack.toEndpoint) ?: return null
+        val nodeDescriptor = trackNodePool[nextNode]
+        val entryPort = nodeDescriptor.getPort(curTrack.toEndpoint)
+        if (entryPort.isNone) return null
```
Comment (bougue-pe) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt]: `null` case is actually an internal error

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt [hunk 15]
```diff
@@ -227,6 +253,20 @@ class RawInfraFromRjsBuilder {
...
+        val exitPort = nodeDescriptor.getTrackNodeExitPort(config, entryPort.asIndex())
+        if (exitPort.isNone) return null
```
Comment (bougue-pe) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/new_impl/RawInfraFromRjsBuilder.kt:265]: `null` case is an invalid  rjsRoute error (handled on call)

**Review by bougue-pe — COMMENTED** [review]
Some comments may be dropped, we should mainly discuss errors and additional comparisons (I'll try to tackle some).

File: core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt (hunk 8 above)
Comment (multun) [reply core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt]: good idea, done

**Pull request merged**